src/multi_van.h

```diff
@@ -0,0 +1,262 @@
+
+// Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+// =============================================================================
+
+#ifndef PS_MULTI_VAN_H_
+#define PS_MULTI_VAN_H_
+
+#include <errno.h>
+#include <fcntl.h>
+#include <netdb.h>
+#include <poll.h>
+#include <stddef.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <zmq.h>
+#include <sys/uio.h>
+
+#include <algorithm>
+#include <map>
+#include <queue>
+#include <set>
+#include <sstream>
+#include <string>
+#include <thread>
+#include <tuple>
+#include <unordered_map>
+#include <vector>
+
+#include "ps/internal/threadsafe_queue.h"
+#include "ps/internal/van.h"
+#include "van_common.h"
+
+namespace ps {
+
+const int ID_OFFSET = 1000000;
+const int MAX_NUM_IDS = 1000;
+
+class MultiVanBufferContext {
+public:
+  int msg_len;
+  int src_idx;
+  Message msg;
+};
+
+class MultiVan : public Van {
+ public:
+  MultiVan() { }
+
+  ~MultiVan() { PS_VLOG(1) << "~MultiVan"; }
+
+  virtual std::string GetType() const {
+    return std::string("multivan");
+  }
+
+ protected:
+  void Start(int customer_id, bool standalone) override {
+    start_mu_.lock();
+    should_stop_ = false;
+
+    auto val = Environment::Get()->find("DMLC_ROLE");
+    std::string role(val);
+    LOG(INFO) << "This is a " << role;
+    start_mu_.unlock();
+
+    // only 1 port needed for the scheduler
+    is_scheduler_ = "scheduler" == role;
+    num_ports_ = 1;
+    const char *npstr = Environment::Get()->find("DMLC_NUM_PORTS");
+    if (npstr && !is_scheduler_) num_ports_ = atoi(npstr);
+
+    // van creation
+    for (int i = 0; i < num_ports_; ++i) {
+      auto van = Van::Create("zmq");
```

Comment:
does it mean that:
1. ps-lite will create separate van for every port and no need to create `sub_vans`?
2. every transport van will represent single local port but connect to all ports of the remote node?

Reply:
Good question. This multi van is added just for testing purpose to verify that sarray device info, and the port info are passed as meta data correctly between nodes. For UCXVan, as we discussed last time, we still expect a single UCXVan instance will handle data from all devices. UCXVan can leverage the ports info added in `Node` to do what it thinks is the best way to handle data from difference devices.